core/node/node_storage_init/src/external_node/snapshot_recovery.rs

```diff
@@ -0,0 +1,82 @@
+use std::{sync::Arc, time::Instant};
+
+use anyhow::Context as _;
+use tokio::sync::watch;
+use zksync_dal::{ConnectionPool, Core};
+use zksync_health_check::AppHealthCheck;
+use zksync_object_store::ObjectStoreFactory;
+use zksync_shared_metrics::{SnapshotRecoveryStage, APP_METRICS};
+use zksync_snapshots_applier::{
+    RecoveryCompletionStatus, SnapshotsApplierConfig, SnapshotsApplierTask,
+};
+use zksync_web3_decl::client::{DynClient, L2};
+
+use crate::{InitializeStorage, SnapshotRecoveryConfig};
+
+#[derive(Debug)]
+pub struct ExternalNodeSnapshotRecovery {
+    pub client: Box<DynClient<L2>>,
+    pub pool: ConnectionPool<Core>,
+    pub recovery_config: SnapshotRecoveryConfig,
+    pub app_health: Arc<AppHealthCheck>,
+}
+
+#[async_trait::async_trait]
+impl InitializeStorage for ExternalNodeSnapshotRecovery {
+    async fn initialize_storage(&self, stop_receiver: watch::Receiver<bool>) -> anyhow::Result<()> {
+        let pool = self.pool.clone();
+        tracing::warn!("Proceeding with snapshot recovery. This is an experimental feature; use at your own risk");
+        let object_store_config =
+            self.recovery_config.object_store_config.clone().context(
+                "Snapshot object store must be presented if snapshot recovery is activated",
+            )?;
+        let object_store = ObjectStoreFactory::new(object_store_config)
+            .create_store()
+            .await?;
```

Comment:
Maybe it makes sense to inject `ObjectStore` directly as a field of `ExternalNodeSnapshotRecovery` instead of using `SnapshotRecoveryConfig.object_store_config` to make it more decoupled from the config?

Reply:
Certainly, but I think the scope of refactoring is much larger (e.g. we need to express different meanings for the same types better, like now we have several kinds of object store, and also several kinds of private key signers), so I think we'll tackle that separately to not over-extend the scope.